Return MediaTime::zeroTime() when m_mediaTimeDuration is invalid at MediaPlayerPrivateGStreamerMSE::duration().

ziransun opened the pull request:
#### 770fbed05b838d191dd40aee6e97c1f83765f2ac
<pre>
Return MediaTime::zeroTime() when m_mediaTimeDuration is invalid at MediaPlayerPrivateGStreamerMSE::duration().
<a href="https://bugs.webkit.org/show_bug.cgi?id=281291">https://bugs.webkit.org/show_bug.cgi?id=281291</a>

Reviewed by Alicia Boya Garcia.

Bug 280337 added an assertion on validation check for currentTime in
ManageMediaSource::monitorSourceBuffers(). During currentTime
calculation, we call MediaPlayerPrivateGStreamerMSE::duration() to work out the
playbackpostion for the MSE case. The initial m_mediaTimeDuration is NaN. The creation
of a element has its currentTime attribute set to MediaTime::zeroTime() though.
This causes the assertion failure when ManageMediaSource is enabled.
Affected test is  media/media-source/media-detachablemse-append.html.

* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::duration const):

Canonical link: <a href="https://commits.webkit.org/285648@main">https://commits.webkit.org/285648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f4923b8a3ce2f6a0e57d2e05a037cc6db8223a42

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73196 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52625 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26004 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77423 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24434 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75311 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60430 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/407 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57524 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15997 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76263 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47542 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62992 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37943 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44181 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20470 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22763 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66035 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20824 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79078 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/724 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65953 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/652 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63001 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65236 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16156 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9135 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7233 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/475 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3212 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Checked out pull request; Running scan-build") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/504 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/489 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/506 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ntrrgc reviewed (approved):
You would think that an unknown duration should be invalid instead of zero, but the WebKit API really expects 0, and 0 is used in other usages, so this is an r+.

ziransun committed:
Return MediaTime::zeroTime() when m_mediaTimeDuration is invalid at MediaPlayerPrivateGStreamerMSE::duration().
https://bugs.webkit.org/show_bug.cgi?id=281291

Reviewed by Alicia Boya Garcia.

Bug 280337 added an assertion on validation check for currentTime in
ManageMediaSource::monitorSourceBuffers(). During currentTime
calculation, we call MediaPlayerPrivateGStreamerMSE::duration() to work out the
playbackpostion for the MSE case. The initial m_mediaTimeDuration is NaN. The creation
of a element has its currentTime attribute set to MediaTime::zeroTime() though.
This causes the assertion failure when ManageMediaSource is enabled.
Affected test is  media/media-source/media-detachablemse-append.html.

* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::duration const):

Canonical link: https://commits.webkit.org/285648@main
